[llmisvc] Update workload and router reconciliation

pierDipi opened the pull request:
- Update single and multi node workload reconciliation to have service accounts where we can attach additional roles
- Improve router-related status conditions and reporting with specific sub-conditions
- Add additional comments to every function and critical logic
- Add envtest-based tests for single, multi node and routing

pierDipi committed:
Update to working reconcile workload

Signed-off-by: Pierangelo Di Pilato <pierdipi@redhat.com>

pierDipi committed:
Add main reconciliation logic in controller.go

Signed-off-by: Pierangelo Di Pilato <pierdipi@redhat.com>

pierDipi committed:
Update router reconciliation

Signed-off-by: Pierangelo Di Pilato <pierdipi@redhat.com>

pierDipi committed:
Add envtest tests

Signed-off-by: Pierangelo Di Pilato <pierdipi@redhat.com>

pierDipi committed:
Add comments to llmisvc controller functions

Signed-off-by: Pierangelo Di Pilato <pierdipi@redhat.com>

pierDipi committed:
Run precommit

Signed-off-by: Pierangelo Di Pilato <pierdipi@redhat.com>

pierDipi committed:
Add test CRDs and update Makefile manifests generation

Signed-off-by: Pierangelo Di Pilato <pierdipi@redhat.com>

pierDipi committed:
Add CRD validation

Signed-off-by: Pierangelo Di Pilato <pierdipi@redhat.com>

yuzisun commented on charts/llmisvc-resources/templates/gateway-inference-extension.yaml:
```diff
@@ -0,0 +1,540 @@
+metadata:
+  annotations:
+    controller-gen.kubebuilder.io/version: v0.16.1
+  name: inferencemodels.inference.networking.x-k8s.io
```
Do we want to install GIE as part of kserve?

pierDipi replied:
I think generally we want to have some control of the versioning of GIE (as Kserve is the primary controller of GIE and EPP), however, we could have a bring your own option

yuzisun commented:
@pierDipi Let me know when this is ready for review

pierDipi committed:
Merge branch 'master' into update-reconcile-workload

pierDipi committed:
Add llmisvc samples

Signed-off-by: Pierangelo Di Pilato <pierdipi@redhat.com>

pierDipi committed:
Propagate network configuration annotations and fix data-parallel-size argument for multi node

Signed-off-by: Pierangelo Di Pilato <pierdipi@redhat.com>

pierDipi committed:
No longer prematurely exit llmisvc route reconciliation

This premature exit prevents the rest of route reconciliation so
the updateRoutingStatus method, which adds the URLs and addresses
associated with llmisvc's routes to the llmisvc spec, never gets
called.

Signed-off-by: Pierangelo Di Pilato <pierdipi@redhat.com>

pierDipi committed:
Fix build - rename controller receiver

Signed-off-by: Pierangelo Di Pilato <pierdipi@redhat.com>

pierDipi committed:
Run make manifests

Signed-off-by: Pierangelo Di Pilato <pierdipi@redhat.com>

terrytangyuan commented on pkg/controller/v1alpha1/llmisvc/scheduler.go:
```diff
@@ -1,5 +1,5 @@
 /*
-Copyright 2025 The KServe Authors.
+Copyright 2023 The KServe Authors.
```
This seems unintentional?

yuzisun commented on pkg/controller/v1alpha1/llmisvc/controller.go:
```diff
@@ -75,84 +88,40 @@ type Config struct {
 	} else {
-		logf.Log.Info("Marked for deletion, finalizing resources")
+		// Resource is being deleted, perform cleanup
+		logger.Info("Marked for deletion, finalizing resources")
```
log the resource name

pierDipi replied:
I will add this to the logger so that it's propagated and used in every log line during reconciliation

yuzisun commented on pkg/controller/v1alpha1/llmisvc/controller.go:
```diff
@@ -163,76 +132,85 @@ func (r *LLMISVCReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ct
 
 	if reconcileErr != nil {
-		logf.Log.Error(reconcileErr, "Failed to reconcile LLMInferenceService")
+		logger.Error(reconcileErr, "Failed to reconcile LLMInferenceService")
```
same here for logging the resource name

yuzisun commented on pkg/controller/v1alpha1/llmisvc/controller.go:
```diff
@@ -163,76 +132,85 @@ func (r *LLMISVCReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ct
 	}
 	llmSvc.MarkPresetsCombinedReady()
+
+	logger.Info("Reconciling with combined base configurations", "combined.spec", baseCfg.Spec, "original.spec", llmSvc.Spec)
```
this should be debug? spec might be large.

yuzisun commented on pkg/controller/v1alpha1/llmisvc/controller.go:
```diff
@@ -248,8 +226,162 @@ func (r *LLMISVCReconciler) updateStatus(ctx context.Context, desired *v1alpha1.
+}
+
+// enqueueOnGatewayChange creates an event handler that triggers reconciliation of LLMInferenceServices
+// when a Gateway they depend on changes. This is necessary because Gateway changes can affect routing.
```
what are the gateway changes what can affect LLMInferenceServices?

pierDipi replied:
The Gateway(s) are fetched to set the llmisvc URL(s) in the status (see router_discovery.go `DiscoverURLs`)

yuzisun commented on pkg/controller/v1alpha1/llmisvc/workload.go:
```diff
@@ -22,51 +22,118 @@ import (
+					Port:        8000,
+					TargetPort: intstr.IntOrString{
+						Type:   intstr.Int,
+						IntVal: 8000,
```
why this is hard coded ? it depends on the workload ports defined on container spec

pierDipi replied:
The rationale for not supporting changing the vLLM port is because it requires a lot of changes beyond this service (presets, routing sidecar flags, routing sidecar container spec , etc) or selecting the exact port is not trivial, for example, when there is P/D, the request goes through the routing sidecar, so here we would need to use the routing sidecar container ports.

The presets we bundle assume `8000` for the "main receiver" (being it vllm or P/D routing sidecar) and for now, I'd discourage that use case unless we have a use case that requires not assuming "main receiver" port.

yuzisun replied:
Ok let’s add a comment then.

pierDipi replied:
Added

yuzisun commented on pkg/controller/v1alpha1/llmisvc/workload_multi_node.go:
```diff
@@ -1,5 +1,5 @@
 /*
-Copyright 2025 The KServe Authors.
+Copyright 2023 The KServe Authors.
```
```suggestion
Copyright 2025 The KServe Authors.
```

yuzisun commented on pkg/controller/v1alpha1/llmisvc/workload_single_node.go:
```diff
@@ -203,24 +241,30 @@ func (r *LLMISVCReconciler) propagateDeploymentStatus(ctx context.Context, expec
 		}
 	}
-
+	logger.Info("Deployment processed")
```
this logging can be removed

pierDipi committed:
Reviews

Signed-off-by: Pierangelo Di Pilato <pierdipi@redhat.com>

pierDipi committed:
Propagate full Config object throughout

Signed-off-by: Pierangelo Di Pilato <pierdipi@redhat.com>

pierDipi committed:
Debug level logging for combined spec

Signed-off-by: Pierangelo Di Pilato <pierdipi@redhat.com>

pierDipi committed:
Add comment for service ports

Signed-off-by: Pierangelo Di Pilato <pierdipi@redhat.com>

yuzisun committed:
Merge branch 'master' into update-reconcile-workload

yuzisun committed:
fix expected hf env secret ValueFrom

Signed-off-by: Dan Sun <dsun20@bloomberg.net>
